### ISSUE TITLE
APIs for helping people unit test with Realm

### ISSUE THREAD

**Issue by austinzheng** [realm/realm-cocoa#3725]
As a follow-up to #3345:

How should we expose APIs or other affordances to help people write correct unit tests that involve Realms?

I talked with Thomas briefly about just exposing `[RLMRealm resetRealmState]` to public users. It turns out that this method is difficult to use correctly (i.e. in a way that won't cause tests to crash nondeterministically). 

We have some [shared test code](https://github.com/realm/realm-cocoa/blob/master/Realm/Tests/RLMTestCase.m#L127) that we use for our own tests. Perhaps we could expose an XCTest-agnostic helper class that abstracts out whatever we need to do to make tests run correctly, and have users use that instead? Strawman example:

```swift
class RealmUnitTestHelper {
    init(setupBlock: RealmTestBlock = nil, teardownBlock: RealmTestBlock = nil)

    // In-memory Realm is named automatically
    func runTest(testBlock: RealmTestBlock)

    func runTestWithRealmNamed(name: String, testBlock: RealmTestBlock)
}

// Usage:
class MyTest : XCTestCase {
    let helper = RealmUnitTestHelper(setupBlock: { ... }, teardownBlock: { ... })

    func testFoo() {
        helper.runTest() { ... }
    }

    func testBar() {
        helper.runTest() { ... }
    }
}
```

Any other ideas are welcome.

@bdash @tgoyne @jpsim @mrackwitz @kishikawakatsumi @stel

**Comment by tgoyne** [comment]
Having to wrap the body of every test in an extra block is somewhat unpleasant: it results in an extra level of indentation for everything, and the extra boilerplate encourages doing more in each test body rather than splitting things into as small of tests as possible. Something along those lines would be to provide a wrapper for `invokeTest()`:

    class MyTest : XCTestCase {
        let helper = RealmUnitTestHelper(setupBlock: { ... }, teardownBlock: { ... })

        override func invokeTest() {
            helper.invokeTest { super.invokeTest() }
        }

        func testDefaultRealm() {
            let realm = Realm() // default realm is set to a per-test file
            // ... do stuff ...
        }

        func testNamedRealm() {
            let realm = helper.realmNamed("named for in memory realm")
            // ... do stuff ...
        }

        func testCustomConfig() {
            let realm = helper.configuationWithName("name")
        }
    }

**Comment by austinzheng** [comment]
I like this idea. I'm wondering if there's a way to enforce the "invokeTest()" part of it, although I think documentation should be sufficient to ensure people use the test helper correctly.

**Comment by mrackwitz** [comment]
I like as well where this is going.

In the best case, this can be exposed in a dedicated `Test` module, defined in the modulemap, to emphasize the scope of the APIs. We might be able to utilize the [link declaration](http://clang.llvm.org/docs/Modules.html#link-declaration) to ensure to conditionally link against XCTest and could use that to expose a ready-to-use `XCTestCase` subclass.
Unfortunately such a selective availability would mean that we can't wrap it in Swift, as that doesn't allow submodules for a single binary yet.

**Comment by austinzheng** [comment]
I agree that our end goal should be a drop-in `XCTestCase` subclass that users can just subclass and use. Swift's tooling and packaging story is probably not good enough yet for this to be viable initially, although we should keep it on our roadmap.

**Comment by austinzheng** [comment]
Yeah, the best case would probably be a full-fledged `XCTest` subclass that a user could just inherit from and use directly. I don't think the tooling is mature enough to make that practica
